index.js

```diff
@@ -797,17 +797,16 @@
   //. the type comprising every [`String`][] value except `''`.
   //.
   //. The given type must satisfy the [Monoid][] and [Setoid][] specifications.
-  var NonEmpty = UnaryType(
-    'sanctuary-def/NonEmpty',
-    functionUrl('NonEmpty'),
-    Any,
-    function(x) {
+  function NonEmpty(parent) {
+    function test(x) {
       return Z.Monoid.test(x) &&
              Z.Setoid.test(x) &&
              !Z.equals(x, Z.empty(x.constructor));
-    },
-    function(monoid) { return [monoid]; }
-  );
+    }
+    function extract(monoid) { return [monoid]; }
+    var t = UnaryTypeWithUrl('sanctuary-def/NonEmpty', parent, test, extract);
```

Comment:
In it's general form, I think the issue could be described as: *When the validation of a refined type fails on its parent type, the reported `propPath` may not correctly correspond to the location of the value as it exists in the refined type*

I think we might even encounter this issue if we refine a record type to be "strict", as has always been the suggestion:

```js
const User = $.RecordType({ name: $.String, password: $.String });
const StrictUser = $.NullaryType('', '', User, x => Object.keys(x).length === User.keys.length);

const validation = StrictUser.validate({ name: 'bob', password: 1 });
// propPath = [ 'password' ]

StrictUser.types[propPath[0]] // => undefined
// the above expression should never return undefined, as the propPath must
// correspond to the types mapping. It does because our propPath is actually
// corresponding to another types mapping!
```